ext/imap/php_imap.c

```diff
@@ -138,28 +140,44 @@ ZEND_TSRMLS_CACHE_DEFINE()
 ZEND_GET_MODULE(imap)
 #endif
 
-/* True globals, no need for thread safety */
-static int le_imap;
+/* Imap class entry definition */
+static zend_class_entry *php_imap_ce = NULL;
+static zend_object_handlers imap_object_handlers;
 
-// TODO Promote to ValueError?
-#define PHP_IMAP_CHECK_MSGNO(msgindex, arg_pos)	\
-	if (msgindex < 1) { \
-		zend_argument_value_error(arg_pos, "must be greater than 0"); \
-		RETURN_THROWS(); \
-	} \
-	if (((unsigned) msgindex) > imap_le_struct->imap_stream->nmsgs) { \
-		php_error_docref(NULL, E_WARNING, "Bad message number"); \
-		RETURN_FALSE;	\
-	}	\
+typedef struct _php_imap_object {
+	MAILSTREAM *imap_stream;
+	long flags;
+	zend_object std;
+} php_imap_object;
 
-/* {{{ mail_close_it */
-static void mail_close_it(zend_resource *rsrc)
-{
-	pils *imap_le_struct = (pils *)rsrc->ptr;
+static inline zend_object *imap_object_to_zend_object(php_imap_object *obj) {
+	return ((zend_object*)(obj + 1)) - 1;
```

Comment:
These pointers are supposed to have the same size, so this could be simplified to `return (zend_object*)obj`. That looks wrong, though. Why not `return obj->std`?

Reply:
NGL, I have no idea, I just copy pasted this from Sara's PR converting (IIRC) FTP resources to object and it was doing that.
But if it can be simplified than I'll do that.

Reply:
For reference about the code I'm talking about: https://github.com/php/php-src/pull/5945/files#diff-526bea3bd44c00ccf1676ead9aeaac72a22e780107a86e0bebe3b81e5c67fb68R66-R73

Reply:
They're not the same size.   sizeof(php_imap_object) is larger than sizeof(zend_object) by definition because the former contains an instance of the latter.

    |-------------- php_imap_object ----------------|
    |-- imap specific data --|-- zend_object data --|

obj+1 takes us from the left edge of the above diagram to the right edge, then the cast followed by the -1 brings us into the middle of it (and thus the start of the zend_object struct).

We *could* do `return &(obj->std);` but when reversing this there's no such convenient expression, so I like to use the math which is visibly reversible in both helpers.

Reply:
Ah, of course! Thanks for the explanation. :)